src/util/psbt/serialize.rs

```diff
@@ -144,3 +155,172 @@ impl Deserialize for SigHashType {
+            .map_err(|_| encode::Error::ParseFailed("invalid control block"))
+    }
+}
+
+// Versioned Script
+impl Serialize for (Script, LeafVersion) {
+    fn serialize(&self) -> Vec<u8> {
+        let mut buf = self.0.to_bytes();
+        buf.push(self.1.as_u8());
+        buf
+    }
+}
+
+impl Deserialize for (Script, LeafVersion) {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        if bytes.is_empty() {
+            return Err(io::Error::from(io::ErrorKind::UnexpectedEof).into())
+        }
+        // The last byte is LeafVersion.
+        let script = Script::deserialize(&bytes[..bytes.len() - 1])?;
+        let leaf_ver = LeafVersion::from_u8(bytes[bytes.len() - 1])
+            .map_err(|_| encode::Error::ParseFailed("invalid control block"))?;
+        Ok((script, leaf_ver))
+    }
+}
+
+
+impl Serialize for (Vec<TapLeafHash>, KeySource) {
+    fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::new();
+        self.0.consensus_encode(&mut buf).expect("Vec's don't error allocation");
+        // TODO: Add support for writing into a writer for key-source
+        buf.extend(self.1.serialize());
+        buf
+    }
+}
+
+impl Deserialize for (Vec<TapLeafHash>, KeySource) {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        let (leafhash_vec, consumed) = deserialize_partial::<Vec::<TapLeafHash>>(&bytes)?;
+        let key_source = KeySource::deserialize(&bytes[consumed..])?;
+        Ok((leafhash_vec, key_source))
+    }
+}
+
+impl Serialize for TapTree {
+    fn serialize(&self) -> Vec<u8> {
+        match (self.0.branch.len(), self.0.branch.last()) {
+            (1, Some(Some(root))) => {
+                let mut buf = Vec::new();
+                for leaf_info in root.leaves.iter() {
+                    // # Panics:
+                    //
+                    // TaprootMerkleBranch can only have len atmost 128(TAPROOT_CONTROL_MAX_NODE_COUNT).
+                    // safe to cast from usize to u8
+                    buf.push(leaf_info.merkle_branch.as_inner().len() as u8);
+                    buf.push(leaf_info.ver.as_u8());
+                    leaf_info.script.consensus_encode(&mut buf).expect("Vec's dont err");
+                }
+                buf
+            }
+        // This should be unreachable if the user directly operates on
+        // psbt API, as we Taptree is already finalized
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Deserialize for TapTree {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        let mut builder = TaprootBuilder::new();
+        let mut read = 0;
+        // Change the psbt traits to non-allocating methods and
+        loop {
+            match (bytes.get(read), bytes.get(read + 1)) {
+                (Some(depth), Some(ver)) => {
+                    let (script, consumed) = deserialize_partial::<Script>(&bytes[read + 2..])?;
+                    read += 2 + consumed;
```

Comment:
Use iterator?

```rust
let mut bytes = bytes.iter();
while let Some(depth) = bytes.next() {
    let version = bytes.next().ok_or(encode::Error::ParseFailed("Invalid Taproot Builder"))?;
    let (script, consumed) = deserialize_partial::<Script>(bytes.as_slice())?;
    if consumed.len() > 0 {
        bytes.nth(consumed.len());
    }

    let leaf_version = LeafVersion::from_u8(*version)
        .map_err(|_| encode::Error::ParseFailed("Leaf Version Error"))?;
    builder = builder.add_leaf_with_ver(usize::from(*depth), script, leaf_version)
        .map_err(|_| encode::Error::ParseFailed("Tree not in DFS order"))?;
}
```

(also notice `usize::from(depth)` instead of `as`)

Reply:
I did try to write this with iterator, but I don't think 
```rust
deserialize_partial::<Script>(bytes.as_slice())?;
```
we can call `as_slice()` from an iterator? This is why I had to do `&[bytes[read + 2..]]`.

Reply:
Thanks, looks like there is https://doc.rust-lang.org/std/slice/struct.Iter.html#method.as_slice . If the iterator was created from a slice, I can call `as_slice()` again on it.

Reply:
This is the same reason why I could not implement the `ControlBlock` derealization using iterators when you initially suggested it.